### ISSUE TITLE
Reproducibility issues in Shipway and Hill example

### ISSUE THREAD

**Issue by jtbuch** [open-atmos/PySDM#1151]
I am trying to reproduce results in the `<..>/Shipway_and_Hill_2012/fig_1.ipynb`  example notebook with no additional changes.  However, there appears to be a clear discrepancy between the outputs on Colab and on my laptop, with the latter showing weird spiky vertical profiles.

For instance, in the rhow=2 and prec= False case,

Colab
![qc_rhow=2_p=False](https://github.com/open-atmos/PySDM/assets/29380821/6cc6b4dd-7f65-4c22-8369-a672e77f86ce)

Laptop
![qc_rhow=2_p=False](https://github.com/open-atmos/PySDM/assets/29380821/260acb3a-9d49-4e50-b18e-e191e20b3389)

Is there any obvious reason why this may be happening? Perhaps a rogue random seed (or precision value) somewhere in the workflow that I can fix to a certain value?

Laptop details:
* Apple M2 Pro
* macOS Ventura 13.4.1 
* python=3.9.16

Edit:

I also find a similar issue in the laptop generated spectrum plot in the `<..>/Yang_et_al_2018/fig_2.ipynb` example notebook as well,

![spectrum](https://github.com/open-atmos/PySDM/assets/29380821/4b52c9c8-7d01-4c1d-b853-d979fd0ba6a8)

**Comment by slayoo** [comment]
@jtbuch Thank you for the report!
I do not recall anyone reporting analogous issues.
The Yang_et_al_2018 example is a condensation-only case, so it narrows the scope of the issue to condensation. Could you check if it also happens with condensation timestep adaptivity turned off? (i.e., passing `adaptive=False` to Condensation dynamic `__init__` method), e.g., here: https://github.com/open-atmos/PySDM/blob/main/examples/PySDM_examples/Yang_et_al_2018/settings.py#L45

**Comment by jtbuch** [comment]
Hi @slayoo, thanks for the prompt response!

After tinkering around with the examples on a couple of different computers, I think the main culprit might be the Apple M1/M2 chips. Especially because I was able to reproduce all your results in a cloud based JupyterHub that runs on the usual ubuntu environment.

**Comment by slayoo** [comment]
Well, this is still a valid unresolved issue then!

**Comment by jtbuch** [comment]
Fair enough! Sorry for prematurely closing it, in that case.

I am actively working on the code, so I will create a PR if I find a fix or another explanation for the issue.

**Comment by jtbuch** [comment]
@slayoo The following `RuntimeWarning` is currently my best guess for the source of the discrepancy (partly because it appears similar to this [issue](https://github.com/onnx/onnx/issues/4774)), but I don't know the `PySDM` code well enough to try to fix it:

```
<..>/site-packages/numpy/core/numeric.py:330: RuntimeWarning: invalid value encountered in cast multiarray.copyto(a, fill_value, casting='unsafe')
```

**Comment by slayoo** [comment]
@jtbuch we've just received a similar report, also from a user of a Apple-silicon machine.

Since despite the noisy values visible in the plot, the simulation seems to continue unaffected, I suspect the problem is with the diagnostics code, like computation of size-spectral statistics (moment products). The errors seem like random noise, but certainly not random numbers, the order of magnitude of the values in the plots still makes some physical sense. One possible culprit could the be the multi-threaded code for computing statistical moments of the spectrum for output, which requires atomic additions that could perhaps be somehow hardware-dependent (and non-atomicity of additions would not introduce garbage values, but would result in omitting some super droplets randomly). If running the code on single thread would help, that would be a confirmation. 

Could you please try running it with NUMBA_NUM_THREADS=1 env var set?

Thanks!

**Comment by pb475** [comment]
Experiencing a similar issue, also on a macbook pro ventura but with the M1 chip, Python3.11.4, PySDM==2.30.post6.

Can confirm that `export NUMBA_NUM_THREADS=1` fixed my problem.

**Comment by slayoo** [comment]
Thank you @pb475 for confirmation!
Stay tuned for a quick bugfix release that will disable the multi-threaded diagnostics for Apple silicon...

**Comment by slayoo** [comment]
(this is likely the root of the problem: https://github.com/open-atmos/PySDM/blob/main/PySDM/backends/impl_numba/atomic_operations.py)

**Comment by slayoo** [comment]
Merged and release, @jtbuch @pb475 could you please check if updating PySDM via `pip install pysdm=2.31` helps with the original problem, and if a warning is show indicating that multi-threading is disabled for ARM64? Thanks!

**Comment by pb475** [comment]
@slayoo, can confirm I get the warning message: 
`UserWarning: Disabling Numba threading due to ARM64 CPU (atomics do not work yet)`

And when I run my original code, all of the apparent 'noise' I was getting (similar to original issue logged here) is now gone.

**Comment by jtbuch** [comment]
@slayoo Sorry for the late response here, but the fix is good for me too!

**Comment by slayoo** [comment]
thank you @jtbuch for confirming!
Disabling multi-threading for M1 is not really a fix, let's track the issue further at #1183
